lib/rspec/matchers/aliased_matcher.rb

```diff
@@ -105,12 +105,53 @@ def failure_message_when_negated
       def optimal_failure_message(same, inverted)
         if DefaultFailureMessages.has_default_failure_messages?(@base_matcher)
           base_message = @base_matcher.__send__(same)
-          overriden    = @description_block.call(base_message)
+          overriden    = @description_block.call(base_message, true)
           return overriden if overriden != base_message
         end
 
         @base_matcher.__send__(inverted)
       end
     end
+
+    # Wrapper around an overriden description for an aliased
+    # matcher.
+    #
+    # @api private
+    class AliasDescription
+      def initialize(new_name, old_name, override_block)
+        @new_name = new_name
+        @old_name = old_name
+        @override_block = override_block
+      end
+
+      # Creates the overriden description.
+      #
+      # @api private
+      def call(old_desc, replace_after_start=false)
+        if @override_block
+          @override_block.call(old_desc)
+        else
+          default_description(old_desc, replace_after_start)
+        end
+      end
+
+    private
+
+      # Generates a default message when no block is provided
+      # when aliasing the matcher. Replaces instances of the old
+      # matcher's name with the new matcher's name where
+      # applicable.
+      #
+      # #api private
+      def default_description(old_desc, replace_after_start)
+        old_name_split = EnglishPhrasing.split_words(@old_name)
+
+        if old_desc.start_with?(old_name_split) || replace_after_start
+          old_desc.sub(old_name_split, EnglishPhrasing.split_words(@new_name))
+        else
+          old_desc
+        end
+      end
+    end
```

Comment:
This is a lot of added complication, and still doesn't remove the global sub in favour of a "start of string only" replace.